<?xml version="1.0"?>

<robot name="robot">

  <!-- Link: chassis -->
  <link name="chassis">
    <pose>0 0 0.5 0 0 0</pose>

    <inertial>
      <mass value="2"/>
      <inertia ixx="0.0033375" ixy="0" ixz="0" iyy="0.0042042" iyz="0" izz="0.0056667"/>
    </inertial>

    <visual name="visual">
      <geometry>
        <box size="0.13 0.12 0.075"/>
      </geometry>
      <material name="blue">
        <color rgba="0.0 0.0 1.0 1"/>
      </material>
    </visual>

    <collision name="collision">
      <geometry>
        <box size="0.14 0.12 0.075"/>
      </geometry>
    </collision>
  </link>

  <link name="left_wheel">
    <inertial>
      <mass value="0.05"/>
      <inertia ixx="2.60416667e-06" ixy="0" ixz="0" iyy="2.60416667e-06" iyz="0" izz="8.68055556e-07"/>
    </inertial>
    <visual name="visual">
      <geometry>
        <cylinder radius="0.035" length="0.02"/>
      </geometry>
      <material name="red">
        <color rgba="1.0 0.0 0.0 1"/>
      </material>
    </visual>
    <collision name="collision">
      <geometry>
        <cylinder radius="0.035" length="0.02"/>
      </geometry>
    </collision>
  </link>

  <!-- Joint: wheel_joint -->
  <joint name="left_wheel_joint" type="continuous">
    <parent link="chassis"/>
    <child link="left_wheel"/>
    <origin xyz="0.045 0.075 -0.025" rpy="-1.5707 0 0"/> <!-- Adjust as needed -->
    <axis xyz="0 0 1"/>
  </joint>

  <link name="right_wheel">
    <inertial>
      <mass value="0.05"/>
      <inertia ixx="2.60416667e-06" ixy="0" ixz="0" iyy="2.60416667e-06" iyz="0" izz="8.68055556e-07"/>
    </inertial>
    <visual name="visual">
      <geometry>
        <cylinder radius="0.035" length="0.02"/>
      </geometry>
      <material name="red">
        <color rgba="1.0 0.0 0.0 1"/>
      </material>
    </visual>
    <collision name="collision">
      <geometry>
        <cylinder radius="0.035" length="0.02"/>
      </geometry>
    </collision>
  </link>

  <joint name="right_wheel_joint" type="continuous">
    <parent link="chassis"/>
    <child link="right_wheel"/>
    <origin xyz="0.045 -0.075 -0.025" rpy="-1.5707 0 0"/> <!-- Adjust as needed -->
    <axis xyz="0 0 1"/>
  </joint>

  <link name="left_rear_wheel">
    <inertial>
      <mass value="0.05"/>
      <inertia ixx="2.60416667e-06" ixy="0" ixz="0" iyy="2.60416667e-06" iyz="0" izz="8.68055556e-07"/>
    </inertial>
    <visual name="visual">
      <geometry>
        <cylinder radius="0.035" length="0.02"/>
      </geometry>
      <material name="red">
        <color rgba="1.0 0.0 0.0 1"/>
      </material>
    </visual>
    <collision name="collision">
      <geometry>
        <cylinder radius="0.035" length="0.02"/>
      </geometry>
    </collision>
  </link>

  <joint name="left_rear_wheel_joint" type="continuous">
    <parent link="chassis"/>
    <child link="left_rear_wheel"/>
    <origin xyz="-0.05 0.075 -0.025" rpy="-1.5707 0 0"/> <!-- Adjust as needed -->
    <axis xyz="0 0 1"/>
  </joint>

  <link name="right_rear_wheel">
    <inertial>
      <mass value="0.05"/>
      <inertia ixx="2.60416667e-06" ixy="0" ixz="0" iyy="2.60416667e-06" iyz="0" izz="8.68055556e-07"/>
    </inertial>
    <visual name="visual">
      <geometry>
        <cylinder radius="0.035" length="0.02"/>
      </geometry>
      <material name="red">
        <color rgba="1.0 0.0 0.0 1"/>
      </material>
    </visual>
    <collision name="collision">
      <geometry>
        <cylinder radius="0.035" length="0.02"/>
      </geometry>
    </collision>
  </link>

  <joint name="right_rear_wheel_joint" type="continuous">
    <parent link="chassis"/>
    <child link="right_rear_wheel"/>
    <origin xyz="-0.05 -0.075 -0.025" rpy="-1.5707 0 0"/> <!-- Adjust as needed -->
    <axis xyz="0 0 1"/>
  </joint>

<gazebo>

  <plugin name='ignition::gazebo::systems::DiffDrive' filename='libignition-gazebo-diff-drive-system.so'>
    <left_joint>left_wheel_joint</left_joint>
    <right_joint>right_wheel_joint</right_joint>
    <wheel_separation>0.15</wheel_separation>
    <wheel_radius>0.035</wheel_radius>

    <odometry_frame>odom</odometry_frame>
    <robot_base_frame>base_link</robot_base_frame>

    <publish_odom>true</publish_odom>
    <publish_odom_tf>true</publish_odom_tf>
    <publish_wheel_tf>true</publish_wheel_tf>
  </plugin>

  <plugin name='ignition::gazebo::systems::DiffDrive' filename='libignition-gazebo-diff-drive-system.so'>
    <left_joint>left_rear_wheel_joint</left_joint>
    <right_joint>right_rear_wheel_joint</right_joint>
    <wheel_separation>0.15</wheel_separation>
    <wheel_radius>0.035</wheel_radius>
    <odom_publish_frequency>1</odom_publish_frequency>
  </plugin>

</gazebo>

  <link name="base_link"/>

  <joint name="base_joint" type="fixed">
    <parent link="base_link"/>
    <child link="chassis"/>
    <origin xyz="0 0 0.1" rpy="0 0 0"/>
  </joint>


  <joint name="base_footprint_joint" type="fixed">
    <parent link="base_link"/>
    <child link="base_footprint"/>
    <origin xyz="0 0 0" rpy="0 0 0"/>
  </joint>

  <link name="base_footprint">
  </link>

  <joint name="laser_joint" type="fixed">
    <parent link="chassis"/>
    <child link="laser_frame"/>
    <origin xyz="0 0 0.0425" rpy="0 0 0"/>
  </joint>

<link name="laser_frame">
    <visual>
        <geometry>
            <cylinder radius="0.01" length="0.005"/>
        </geometry>
        <material name="red">
          <color rgba="1.0 0.0 0.0 1"/>
        </material>
    </visual>
    <collision>
      <geometry>
        <cylinder radius="0.01" length="0.005"/>
      </geometry>
    </collision>
</link>

<gazebo reference="laser_frame">
    <material>Gazebo/Red</material>

    <sensor name='gpu_lidar' type='gpu_lidar'>
      <topic>simulator_lidar</topic>
      <update_rate>20</update_rate>
      <lidar>
        <scan>
          <horizontal>
            <samples>360</samples>
            <resolution>1</resolution>
            <min_angle>-3.14</min_angle>
            <max_angle>3.14</max_angle>
          </horizontal>

          <vertical>
            <samples>1</samples>
            <resolution>1</resolution>
            <min_angle>0</min_angle>
            <max_angle>0</max_angle>
          </vertical>
        </scan>
        <range>
          <min>0.08</min>
          <max>10</max>
          <resolution>0.01</resolution>
        </range>
        <noise>
          <type>none</type>
          <mean>0</mean>
          <stddev>0</stddev>
        </noise>
        <visibility_mask>4294967295</visibility_mask>
      </lidar>
    </sensor>

</gazebo>

<gazebo>
  <plugin name="ignition::gazebo::systems::JointStatePublisher" filename="libignition-gazebo-joint-state-publisher-system.so">

    <update_rate>10</update_rate>
      <joint_name>left_wheel_joint</joint_name>
      <joint_name>right_wheel_joint</joint_name>
      <joint_name>left_rear_wheel_joint</joint_name>
      <joint_name>right_rear_wheel_joint</joint_name>
      <topic>joint_states</topic>
  </plugin>

</gazebo>


<link name="ultrasonic">
  <visual>
    <geometry>
      <box size="0.01 0.03 0.015"/>
    </geometry>
      <material name="red">
        <color rgba="1.0 0.0 0.0 1"/>
      </material>
  </visual>
  <collision>
    <geometry>
      <box size="0.01 0.03 0.015"/>
    </geometry>
  </collision>
</link>

<joint name="ultrasonic_joint" type="fixed">
  <parent link="chassis"/>
  <child link="ultrasonic"/>
  <origin xyz="0.065 0 -0.02" rpy="0 0 0"/>
</joint>

<gazebo reference="ultrasonic">
  <material>Gazebo/Red</material>

  <sensor name='gpu_lidar_ultra' type='gpu_lidar'>
    <topic>simulator_ultrasonic</topic>
    <update_rate>20</update_rate>
    <lidar>
      <scan>
        <horizontal>
          <samples>3</samples>
          <resolution>1</resolution>
          <min_angle>-0.1308</min_angle>
          <max_angle>0.1308</max_angle>
        </horizontal>
        <vertical>
          <samples>1</samples>
          <resolution>1</resolution>
          <min_angle>0</min_angle>
          <max_angle>0</max_angle>
        </vertical>
      </scan>
      <range>
        <min>0.02</min>
        <max>4</max>
        <resolution>0.01</resolution>
      </range>
      <noise>
        <type>none</type>
        <mean>0</mean>
        <stddev>0</stddev>
      </noise>
      <visibility_mask>4294967295</visibility_mask>
    </lidar>
  </sensor>

</gazebo>


<link name="camera">
  <inertial>
    <mass value="0.01"/>
    <inertia ixx="2.4375e-06" ixy="0" ixz="0" iyy="1.084e-06" iyz="0" izz="2.0208e-06"/>
  </inertial>
  <visual>
    <geometry>
      <box size="0.02 0.045 0.03"/>
    </geometry>
      <material name="red">
        <color rgba="1.0 0.0 0.0 1"/>
      </material>
  </visual>
  <collision>
    <geometry>
      <box size="0.02 0.045 0.03"/>
    </geometry>
  </collision>
</link>

<gazebo reference="camera">
  <sensor name='camera' type='camera'>
    <topic>simulator_camera</topic>
    <update_rate>30</update_rate>
    <camera name='__default__'>
        <horizontal_fov>1.0469999999999999</horizontal_fov>
        <image>
            <width>960</width>
            <height>540</height>
            <format>RGB_INT8</format>
        </image>
        <clip>
            <near>0.01</near>
            <far>100</far>
        </clip>
        <visibility_mask>4294967295</visibility_mask>
        <lens>
            <type>stereographic</type>
            <scale_to_hfov>true</scale_to_hfov>
            <cutoff_angle>1.5708</cutoff_angle>
            <env_texture_size>256</env_texture_size>
        </lens>
    </camera>
</sensor>

</gazebo>

<joint name="camera_servo_joint" type="revolute">
  <parent link="chassis"/>
  <child link="camera"/>
  <origin xyz="0.06 0 0.03" rpy="0 0 0"/>
  <axis xyz="0 1 0"/>
  <limit lower="-3.14" upper="3.14" effort="10" velocity="3.14"/>
</joint>

<gazebo>
  <plugin name="ignition::gazebo::systems::JointController" filename="libignition-gazebo-joint-controller-system.so">
    <joint_name>camera_servo_joint</joint_name>
</plugin>
</gazebo>



<link name="line_tracking_left">
  <visual>
    <geometry>
      <box size="0.01 0.01 0.01"/>
    </geometry>
      <material name="red">
        <color rgba="1.0 0.0 0.0 1"/>
      </material>
  </visual>
  <collision>
    <geometry>
      <box size="0.01 0.01 0.01"/>
    </geometry>
  </collision>
</link>

<gazebo reference="line_tracking_left">
  <sensor name='line_tracking_left' type='camera'>
    <topic>line_tracking</topic>
    <update_rate>10</update_rate>
    <camera name='__default__'>
        <horizontal_fov>0.11</horizontal_fov>
        <image>
            <width>1</width>
            <height>1</height>
            <format>RGB_INT8</format>
        </image>
        <clip>
            <near>0.01</near>
            <far>100</far>
        </clip>
    </camera>
</sensor>

</gazebo>

<joint name="line_tracking_left_joint" type="fixed">
  <parent link="chassis"/>
  <child link="line_tracking_left"/>
  <origin xyz="0.055 0.03 -0.0375" rpy="0 1.57 0"/>
</joint>


<link name="line_tracking_middle">
  <visual>
    <geometry>
      <box size="0.01 0.01 0.01"/>
    </geometry>
      <material name="red">
        <color rgba="1.0 0.0 0.0 1"/>
      </material>
  </visual>
  <collision>
    <geometry>
      <box size="0.01 0.01 0.01"/>
    </geometry>
  </collision>
</link>

<gazebo reference="line_tracking_middle">
  <sensor name='line_tracking_middle' type='camera'>
    <topic>line_tracking</topic>
    <update_rate>10</update_rate>
    <camera name='__default__'>
        <horizontal_fov>0.11</horizontal_fov>
        <image>
            <width>1</width>
            <height>1</height>
            <format>RGB_INT8</format>
        </image>
        <clip>
            <near>0.01</near>
            <far>100</far>
        </clip>
    </camera>
</sensor>

</gazebo>

<joint name="line_tracking_middle_joint" type="fixed">
  <parent link="chassis"/>
  <child link="line_tracking_middle"/>
  <origin xyz="0.055 0 -0.0375" rpy="0 1.57 0"/>
</joint>


<link name="line_tracking_right">
  <visual>
    <geometry>
      <box size="0.01 0.01 0.01"/>
    </geometry>
      <material name="red">
        <color rgba="1.0 0.0 0.0 1"/>
      </material>
  </visual>
  <collision>
    <geometry>
      <box size="0.01 0.01 0.01"/>
    </geometry>
  </collision>
</link>

<gazebo reference="line_tracking_right">
  <sensor name='line_tracking_right' type='camera'>
    <topic>line_tracking</topic>
    <update_rate>10</update_rate>
    <camera name='__default__'>
        <horizontal_fov>0.11</horizontal_fov>
        <image>
            <width>1</width>
            <height>1</height>
            <format>RGB_INT8</format>
        </image>
        <clip>
            <near>0.01</near>
            <far>100</far>
        </clip>
    </camera>
</sensor>

</gazebo>

<joint name="line_tracking_right_joint" type="fixed">
  <parent link="chassis"/>
  <child link="line_tracking_right"/>
  <origin xyz="0.055 -0.03 -0.0375" rpy="0 1.57 0"/>
</joint>
</robot>
